<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Alignment and counting &mdash; Proj029 - Microbiome and host profiling in a mouse model of colitis 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Proj029 - Microbiome and host profiling in a mouse model of colitis 1.0 documentation" href="index.html" />
    <link rel="next" title="MetagenomeSeq initial normalisation" href="metagenomeseq.html" />
    <link rel="prev" title="Introduction" href="introduction.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="metagenomeseq.html" title="MetagenomeSeq initial normalisation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="introduction.html" title="Introduction"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Proj029 - Microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="alignment-and-counting">
<h1>Alignment and counting<a class="headerlink" href="#alignment-and-counting" title="Permalink to this headline">¶</a></h1>
<p>We pick up the analysis at the point at which raw fastq files have
been filtered for contaminating adapters, pairs have been flashed and reads
mapping to rRNA and mouse have been removed. Details of these steps are in the
paper. Both raw fastq files and processed fastq files are available at the EBI ENA
under accession number EMTAB-XXXX so the analysis can be run without having to
perform the pre-processing steps yourself:</p>
<div class="highlight-python"><div class="highlight"><pre>NOTE: The alignment and counting steps can take a long time and large amount of memory. If
you do not want to run these steps then we have provided the counts tables for genera
and functional groups (NOGs) - see bottom of this page.
</pre></div>
</div>
<p>The RNA and DNA analyses were run separately using the same pipeline. For the our
purposes we will run the analyses separately in two separate directories. Create these
directories:</p>
<div class="highlight-python"><div class="highlight"><pre>$ mkdir RNA
$ mkdir DNA
</pre></div>
</div>
<p>Now download the .fastq files from the EBI ENA into their respective directories.</p>
<div class="section" id="alignment">
<h2>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h2>
<p>The first step in the analysis is to assign reads to taxa and functional groups.
In order to do this we need some databases. We have used the NCBI non-redundant
protein database (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz">nr</a>) for taxonomic purposes and the integrated gene catalogue
(<a class="reference external" href="ftp://climb.genomics.cn/pub/10.5524/100001_101000/100064/1.GeneCatalogs/IGC.pep.gz">IGC</a>) for assessing functions. Download these into a databases directory:</p>
<div class="highlight-python"><div class="highlight"><pre>$ mkdir databases
$ cd databases
$ wget ftp://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz
$ wget ftp://climb.genomics.cn/pub/10.5524/100001_101000/100064/1.GeneCatalogs/IGC.pep.gz
</pre></div>
</div>
<p>We used DIAMOND to align sequences to the reference databases so first we build
database indexes. In the databases directory do:</p>
<div class="highlight-python"><div class="highlight"><pre>$ diamond makedb --db nr --in nr --threads 16
</pre></div>
</div>
<p>and:</p>
<div class="highlight-python"><div class="highlight"><pre>$ diamond makedb --db igc --in IGC.pep --threads 16
</pre></div>
</div>
<p>NOTE: DIAMOND takes uncompressed files as input so you will need to unzip first</p>
<p>Next we ran the alignment for each fastq file (reads) to each database
(fastq files need to be converted to fasta format first e.g. using <a class="reference external" href="http://hannonlab.cshl.edu/fastx_toolkit/">fastx toolkit</a>). As
an example:</p>
<blockquote>
<div># change into the RNA directory (contains raw data) for example
$ cd RNA</div></blockquote>
<p>conversion:</p>
<div class="highlight-python"><div class="highlight"><pre>$ fastq-to-fasta stool-HhaIL10R-R1.fastq.gz &gt; stool-HhaIL10R-R1.fasta
</pre></div>
</div>
<p>alignment:</p>
<div class="highlight-python"><div class="highlight"><pre>$ diamond blastx --db ../databases/nr --query stool-HhaIL10R-R1.fasta -o stool-HhaIL10R-R1.diamond.tsv
</pre></div>
</div>
<p>This was done for each sample (DNA-seq and RNA-seq) against each database, resulting in</p>
<p>16x DNA-seq to nr</p>
<p>16x RNA-seq to nr</p>
<p>16x DNA-seq to IGC</p>
<p>16x RNA-seq to IGC</p>
</div>
<div class="section" id="counting">
<h2>Counting<a class="headerlink" href="#counting" title="Permalink to this headline">¶</a></h2>
<p>Our main objective was to identify community structure and functional changes between mice
with and without colitis. In order to do that we first needed to count the number of
reads that mapped to each genus and each functional category. This is where the analysis
of taxonomy and function diverge.</p>
<p>For taxonomic profiling we used the lowest common ancestor approach (LCA) to
assign reads to genera implemented using lcamapper.sh from mtools (see dependencies). This
requires the mapping file of gi number to taxonomy id that is distributed with mtools. For each
sample for both DNA and RNA data sets aligned to nr we do for example:</p>
<div class="highlight-python"><div class="highlight"><pre>$ lcamapper.sh -i stool-HhaIL10R-R1.diamond.tsv -f Detect -ms 50 -me 0.01 -tp 50 -gt gi_taxid_prot.bin -o stool-HhaIL10R-R1.lca
</pre></div>
</div>
<p>Again each file that was aligned to the ncbi nr database is used as input to lcamapper.sh. To obtain counts per
genus we use the lca2table.py script that is in the scripts/ directory
of the <a class="reference external" href="https://github.com/CGATOxford/cgat">CGATOxford/cgat</a> repository:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cat stool-HhaIL10R-R1.lca | python cgat/scripts/lca2table.py --summarise=taxa-counts --log=stool-HhaIL10R-R1.lca.counts.tsv.log &gt; stool-HhaIL10R-R1.lca.counts.tsv
</pre></div>
</div>
<p>At this point each file that contains taxa counts is loaded into an sqlite database. This makes subsetting etc easier
downstream. We use the c2v2db.py script from cgat/ respository to do this:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cat stool-HhaIL10R-R1.counts.tsv | python &lt;path_to_cgat&gt;/cgat/scripts/csv2db.py
                                     --backend=sqlite
                                     --retry
                                     --table=stool_HhaIL10R_R1_lca_counts
                                     &gt; stool-HhaIL10R-R1.lca.counts.tsv.load
</pre></div>
</div>
<p>This will create a database called &#8220;csvdb&#8221; in the working directory and will have loaded the table
stool_HhaIL10R_R1_lca_counts. In our analyses we were interested in genus abundances and so we create
flat files from the database of genus counts by:</p>
<div class="highlight-python"><div class="highlight"><pre>$ sqlite3 csvdb &#39;SELECT taxa, count FROM stool_HhaIL10R_R1_lca_counts WHERE taxa == &quot;genus&quot;&#39; &gt; stool-HhaIL10R-R1.lca.counts.tsv
</pre></div>
</div>
<p>Next we combine counts for each sample into a single table with rows as genera and samples as columns. We use a convenient
script from CGAT code collection to do this - combine_tables.py.</p>
<p>e.g for example if we are in the RNA analysis working directory we run combine_tables.py by specifying that we want missing
genera from a sample to be given a 0 count, we want to merge on column 1 (genus), we want to take the &#8220;count&#8221;
column and we want to combine all tables that end in .lca.counts.tsv. We also specify the prefixes to be used
for each column in the resulting combined table:</p>
<div class="highlight-python"><div class="highlight"><pre>$ python cgat/scripts/combine_tables.py
         --missing=0
         --columns=1
         --take=count
         --glob=*.lca.counts.tsv
         --prefixes=stool-HhaIL10R-R4,
                    stool-HhaIL10R-R3,
                    stool-Hh-R4,
                    stool-Hh-R3,
                    stool-WT-R4,
                    stool-aIL10R-R1,
                    stool-WT-R3,
                    stool-WT-R2,
                    stool-aIL10R-R4,
                    stool-Hh-R2,
                    stool-Hh-R1,
                    stool-aIL10R-R2,
                    stool-WT-R1,
                    stool-HhaIL10R-R1,
                    stool-HhaIL10R-R2,
                    stool-aIL10R-R3
 | gzip &gt; genus.diamond.aggregated.counts.tsv.gz
</pre></div>
</div>
<p>Again we do this for both RNA and DNA data sets. This produces a table (truncated for visual reasons)</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="16%" />
<col width="13%" />
<col width="13%" />
<col width="18%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">taxa</th>
<th class="head">stool-WT-R1_count</th>
<th class="head">stool-WT-R3_count</th>
<th class="head">stool-aIL10R-R1_count</th>
<th class="head">stool-Hh-R2_count</th>
<th class="head">stool-Hh-R1_count</th>
<th class="head">stool-HhaIL10R-R4_count</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Methylobacillus</td>
<td>228</td>
<td>517</td>
<td>560</td>
<td>406</td>
<td>201</td>
<td>353</td>
</tr>
<tr class="row-odd"><td>Methanosphaera</td>
<td>98</td>
<td>224</td>
<td>194</td>
<td>175</td>
<td>65</td>
<td>132</td>
</tr>
<tr class="row-even"><td>Desulfarculus</td>
<td>3</td>
<td>6</td>
<td>12</td>
<td>2</td>
<td>5</td>
<td>2</td>
</tr>
<tr class="row-odd"><td>Polaromonas</td>
<td>859</td>
<td>2021</td>
<td>2034</td>
<td>1111</td>
<td>616</td>
<td>1806</td>
</tr>
<tr class="row-even"><td>Caldanaerobacter</td>
<td>3330</td>
<td>5367</td>
<td>5847</td>
<td>3645</td>
<td>2072</td>
<td>6571</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>The next task is to produce a counts table similar to the one above but for functions. We have aligned to the IGC and we use
their annotations of eggNOG functions (NOGs) as the features to be counted. For counting per NOG, we extract the best hit DIAMOMD alignment
for each read, map the gene to NOG using an additional mapping file (provided in data/ directory) and perform the counting.
To produce a count table for a single sample we use the diamond2counts.py script in the scripts/ directory. The input is the DIAMOND alignment file.</p>
<div class="highlight-python"><div class="highlight"><pre>$ zcat stool-HhaIL10R-R1.igc.tsv.gz | python &lt;path_to_cgat&gt;/cgat/scripts/diamond2counts.py
                                      --method=best
                                      --cog-map=data/gene2cog.tsv.gz
                                      --sum-cog
                                      --log=stool-HhaIL10R-R1.igc.counts.tsv.gz.log
                                    | gzip &gt; stool-HhaIL10R-R1.nogs.counts.tsv.gz
</pre></div>
</div>
<p>Again, we combine tables for each sample as for genera into a final counts table using combine_tables.py to give</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="17%" />
<col width="17%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="16%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">ref</th>
<th class="head">stool-HhaIL10R-R4_count</th>
<th class="head">stool-HhaIL10R-R3_count</th>
<th class="head">stool-Hh-R4_count</th>
<th class="head">stool-Hh-R3_count</th>
<th class="head">stool-WT-R4_count</th>
<th class="head">stool-aIL10R-R1_count</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>NOG243840</td>
<td>2</td>
<td>4</td>
<td>6</td>
<td>10</td>
<td>1</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>NOG281778</td>
<td>1</td>
<td>5</td>
<td>4</td>
<td>4</td>
<td>2</td>
<td>1</td>
</tr>
<tr class="row-even"><td>NOG41625</td>
<td>113</td>
<td>744</td>
<td>414</td>
<td>1273</td>
<td>404</td>
<td>567</td>
</tr>
<tr class="row-odd"><td>bactNOG18808</td>
<td>1</td>
<td>2</td>
<td>8</td>
<td>8</td>
<td>15</td>
<td>21</td>
</tr>
<tr class="row-even"><td>COG3010</td>
<td>2118</td>
<td>2395</td>
<td>1061</td>
<td>1738</td>
<td>2483</td>
<td>1043</td>
</tr>
<tr class="row-odd"><td>proNOG56664</td>
<td>7</td>
<td>2</td>
<td>3</td>
<td>2</td>
<td>4</td>
<td>5</td>
</tr>
<tr class="row-even"><td>COG3012</td>
<td>15</td>
<td>60</td>
<td>34</td>
<td>101</td>
<td>38</td>
<td>22</td>
</tr>
<tr class="row-odd"><td>COG3014</td>
<td>27</td>
<td>18</td>
<td>32</td>
<td>92</td>
<td>41</td>
<td>17</td>
</tr>
<tr class="row-even"><td>COG3015</td>
<td>13</td>
<td>298</td>
<td>91</td>
<td>206</td>
<td>73</td>
<td>1148</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Counts tables for genera and NOGs (DNA and RNA) will now be taken forward to the analysis steps.</p>
<p>Given the large sizes of raw and alignment files these have been deposited at the EBI ENA (ADD LINK). We expect that you will not
run the above steps because it is time-cosuming. Therefore, for reproducing our downstream analysis
we have provided the count tables in the data/DNA/ and data/RNA/ directories (genus.diamond.aggregated.counts.tsv.gz and gene_counts.tsv.gz).</p>
<p>In this case link to the counts tables we have provided and load into the csvdb database.</p>
<p>For RNA:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cd RNA
$ ln -s ../data/RNA/genus.diamond.aggregated.counts.tsv.gz .
$ ln -s ../data/RNA/gene_counts.tsv.gz .
$ zcat genus.diamond.aggregated.counts.tsv.gz | python cgat/scripts/csv2db.py
                                                --backend=sqlite
                                                --retry
                                                --table=genus_diamond_aggregated_counts
                                                &gt; genus.diamond.aggregated.counts.tsv.gz.load

$ zcat gene_counts.tsv.gz | python cgat/scripts/csv2db.py
                            --backend=sqlite
                            --retry
                            --table=gene_counts
                            &gt; gene_counts.tsv.gz.load
</pre></div>
</div>
<p>For DNA:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cd DNA
$ ln -s ../data/DNA/genus.diamond.aggregated.counts.tsv.gz .
$ ln -s ../data/DNA/gene_counts.tsv.gz .
$ zcat genus.diamond.aggregated.counts.tsv.gz | python cgat/scripts/csv2db.py
                                                --backend=sqlite
                                                --retry
                                                --table=genus_diamond_aggregated_counts
                                                &gt; genus.diamond.aggregated.counts.tsv.gz.load


$ zcat gene_counts.tsv.gz | python cgat/scripts/csv2db.py
                            --backend=sqlite
                            --retry
                            --table=gene_counts
                            &gt; gene_counts.tsv.gz.load
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Alignment and counting</a><ul>
<li><a class="reference internal" href="#alignment">Alignment</a></li>
<li><a class="reference internal" href="#counting">Counting</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="introduction.html"
                        title="previous chapter">Introduction</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="metagenomeseq.html"
                        title="next chapter">MetagenomeSeq initial normalisation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/alignmentAndCounting.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="metagenomeseq.html" title="MetagenomeSeq initial normalisation"
             >next</a> |</li>
        <li class="right" >
          <a href="introduction.html" title="Introduction"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Proj029 - Microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Nick Ilott.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b3.
    </div>
  </body>
</html>