<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Defining colitis-dependent NOGs &mdash; Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation" href="index.html" />
    <link rel="next" title="Code" href="code.html" />
    <link rel="prev" title="Principle components analysis" href="pca.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="code.html" title="Code"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pca.html" title="Principle components analysis"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="defining-colitis-dependent-nogs">
<h1>Defining colitis-dependent NOGs<a class="headerlink" href="#defining-colitis-dependent-nogs" title="Permalink to this headline">Â¶</a></h1>
<p>Create a link to the goi.tsv file in the data/ directory. This will become clearer in a little while:</p>
<div class="highlight-python"><div class="highlight"><pre>$ ln -s ../data/goi.tsv .
</pre></div>
</div>
<p>We can run the overlap analyses and PCA for NOGs as we did for genera (DNA) in the previous sections (in the compare_datasets/ directory).
We won&#8217;t therefore go over the same functions.</p>
<p>Out metagenomeSeq analysis was performed on NOGs and we identified 1221 and 669 NOGs to be differentially abundant in
metagenomic and metatranscriptomic data sets, respectively. The overlap between these sets suggested that there were
some NOGs that were transcriptionally activated in colitis, leading us to hypothesise that these were due to responses
by bacteria to an alteration in the gut niche.</p>
<p>A few tasks were performed before we produced Fig. 2g - some of which did not end up in the manuscript. For example we
built the RNA fold / DNA fold ratio for NOGs using the metagenomeSeq differential abundance files. In the comparison.dir:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;&gt; import Proj029Pipelines.PipelineMetaomics as PipelineMetaomics
&gt;&gt; PipelineMetaomics.buildRNADNARatio(&quot;../DNA/gene_counts.diff.tsv&quot;,
                                      &quot;../RNA/gene_counts.diff.tsv&quot;,
                                      &quot;rna_dna_ratio.tsv&quot;)
</pre></div>
</div>
<p>Then we built a list of NOGs that were common between the DNA and RNA data sets from the databases that contain
tables of differential expression results. First load the tables:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;&gt; PipelineMetaomics.buildCommonList(&quot;../RNA/csvdb&quot;,
                                         &quot;../DNA/csvdb&quot;,
                                         &quot;common_genes.tsv&quot;)
</pre></div>
</div>
<p>then we build the differentially abundant NOGs list from database (Hh + aIL10R vs. Steady state):</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;&gt; PipelineMetaomics.buildGeneDiffList(&quot;../RNA/csvdb&quot;,
                                       &quot;common_genes.tsv&quot;,
                                       &quot;rna_diff_genes.tsv&quot;)

&gt;&gt; PipelineMetaomics.buildGeneDiffList(&quot;../DNA/csvdb&quot;,
                                       &quot;common_genes.tsv&quot;,
                                       &quot;dna_diff_genes.tsv&quot;)
</pre></div>
</div>
<p>and just before plotting we annotate NOGs with their differential abundance status i.e
in DNA, RNA or both data sets:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;&gt; PipelineMetaomics.annotateRNADNARatio(&quot;rna_dna_ratio.tsv&quot;,
                                         &quot;dna_diff_genes.tsv&quot;,
</pre></div>
</div>
<dl class="docutils">
<dt>1                                             &#8220;rna_diff_genes.tsv&#8221;,</dt>
<dd>&#8220;rna_dna_ratio.annotated.tsv&#8221;)</dd>
</dl>
<p>and we plot the DNA and RNA fold changes, fit a linear model and plot the 95% prediction intervals. There is
a little bit of sneakyness here as before we created the final plot we found NOGs of interest. This function
picks up a file called goi.tsv that contains annotations for the plot. These annotations are for NOGs that
are annotated as being involved in oxidative stress resistance and are up-regulated in metatranscriptomic
data and defined as colitis-responsive (lying outside of the 95% prediction interval for the model fit):</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;&gt; PipelineMetaomics.plotSets(&quot;rna_dna_ratio.annotated.tsv&quot;,
                              &quot;rna_dna_ratio.annotated.png&quot;)
</pre></div>
</div>
<p>This will produce Fig.2g.</p>
<a class="reference internal image-reference" href="_images/rna_dna_ratio.annotated.png"><img alt="_images/rna_dna_ratio.annotated.png" class="align-center" src="_images/rna_dna_ratio.annotated.png" style="width: 400pt; height: 400pt;" /></a>
<p>And that is Fig. 2 done!</p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="pca.html"
                        title="previous chapter">Principle components analysis</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="code.html"
                        title="next chapter">Code</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/colitisResponsiveNOGs.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="code.html" title="Code"
             >next</a> |</li>
        <li class="right" >
          <a href="pca.html" title="Principle components analysis"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Nick Ilott.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b3.
    </div>
  </body>
</html>