<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>MetagenomeSeq filtered analysis &mdash; Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation" href="index.html" />
    <link rel="next" title="Principle components analysis" href="pca.html" />
    <link rel="prev" title="Comparing metagenomic and metatranscriptomic data sets" href="compareDatasets.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pca.html" title="Principle components analysis"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="compareDatasets.html" title="Comparing metagenomic and metatranscriptomic data sets"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="metagenomeseq-filtered-analysis">
<h1>MetagenomeSeq filtered analysis<a class="headerlink" href="#metagenomeseq-filtered-analysis" title="Permalink to this headline">Â¶</a></h1>
<p>Now we have compared the RNA and DNA data sets and have sets of features that we want to analyse,
we can run metagenomeSeq on a filtered set of features (using common_genera.tsv and common_genes.tsv).</p>
<p>For example to produce new counts tables with filtered features for genera we can do:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cd RNA
$ zcat genus.diamond.aggregated.counts.tsv.gz | python proj029/scripts/counts2restrictedcounts.py
                                              --restrict-to=../compare_datasets/common_genera.tsv
                                              --log=restrict_genera.log
                                              &gt; genus.diamond.aggregated.counts.restricted.tsv
</pre></div>
</div>
<p>We can then use this filtered counts table for metagenomeSeq analysis:</p>
<div class="highlight-python"><div class="highlight"><pre>$ Rscript cgat/R/run_metagenomeseq.R --k 4 --a 0.1 -c genus.diamond.aggregated.counts.restricted.tsv -p genus.diamond.aggregated.counts
</pre></div>
</div>
<dl class="docutils">
<dt>::</dt>
<dd>NOTE: here we have overwritten the previous results tables from metagenomeSeq.</dd>
</dl>
<p>The resulting .norm.matrix files are used in subsequent principle components analysis and the .diff.tsv files are the final
results of differential abundance testing.</p>
<p>Load the differential abundance table into the database:</p>
<div class="highlight-python"><div class="highlight"><pre>$ cat genus.diamond.aggregated.counts.diff.tsv |
  python &lt;path_to_cgat&gt;/cgat/scripts/csv2db.py --backend=sqlite
                                               --retry
                                               --table=genus_diamond_aggregated_counts_diff
                                               &gt; genus.diamond.aggregated.counts.diff.tsv.load
</pre></div>
</div>
<dl class="docutils">
<dt>::</dt>
<dd>NOTE: here we have overwritten the database tables</dd>
</dl>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="compareDatasets.html"
                        title="previous chapter">Comparing metagenomic and metatranscriptomic data sets</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pca.html"
                        title="next chapter">Principle components analysis</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/metagenomeseq_restrict.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pca.html" title="Principle components analysis"
             >next</a> |</li>
        <li class="right" >
          <a href="compareDatasets.html" title="Comparing metagenomic and metatranscriptomic data sets"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Combined microbiome and host profiling in a mouse model of colitis 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Nick Ilott.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b3.
    </div>
  </body>
</html>